### DISCUSSION TITLE
File manager custom action to follow symbolic link

### DISCUSSION THREAD

**Discussion by romildo** [lxqt/lxqt#2501] [Q&A]
I am looking for a custom action for pcmanfm-qt to follow a symlink. When a file (or directory) which is a symbolic link is selected, I want the action to open the directory containing the original file (or directory), and select the original file (or directory).

I would write it myself, but it does not seem so straightforward.

Maybe someone already have this action implemented and want to share it, or want to give guidelines on how to accomplish it.

**Comment by tsujan** [comment]
The target of a symlink can be found by `readlink -f`.

**Comment by romildo** [comment]
Is it possible to open pcmanfm-qt with a given file path, and automatically select the file name in its parent folder?

Something like what is offered by the `--select` option of nautilus and dolphin:

```
$ nautilus --select /var/log/Xorg.0.log
```

![image](https://github.com/lxqt/lxqt/assets/1217934/11d31655-09c3-4eda-889e-17f255a70db1)

```
$ dolphin --select /var/log/Xorg.0.log
```

![image](https://github.com/lxqt/lxqt/assets/1217934/760c8c00-b085-4d66-a813-f3333fc1f3c5)


Thunar is also able to do this:

```
$ thunar /var/tmp/nix-build.log
```

![image](https://github.com/lxqt/lxqt/assets/1217934/aacb116b-72d4-442e-9894-1fc5d2fb937e)

**Reply by tsujan** [reply]
PCManFM-Qt doesn't have a `select` option, but since it supports `org.freedesktop.FileManager1`, you can do it in a generic way. Various apps, like Firefox, Chromium,…, use the generic D-BUS methods to show files/folders or open File Properties dialogs in the currently active file manager. Those methods can be called in a terminal or Bash script too.

When PCManFM-Qt's process is running (e.g., in the desktop or daemon mode), if you're familiar with QDbusViewer, you could find the needed commands easily. Otherwise, see https://github.com/lxqt/pcmanfm-qt/pull/1525#issuecomment-1031619749.

For example, for a single file:

```
qdbus org.freedesktop.FileManager1 /org/freedesktop/FileManager1 ShowItems 'FILE_PATH' "0"
```

And for multiple files:

```
qdbus org.freedesktop.FileManager1 /org/freedesktop/FileManager1 ShowItems "(" 'FILE1_PATH' 'FILE2_PATH' ")" "0"
```

**Comment by romildo** [comment]
I have written the following file manager custom action to follow symbolic links in the file manager:

`~/.local/share/file-manager/actions/follow-symlink.desktop`:
```desktop
[Desktop Entry]
Type=Action
Name=Follow Symbolic Links
Tooltip=Select target files in parent folder
Icon=link
Profiles=profile-zero;

[X-Action-Profile profile-zero]
Exec=follow_symlink.sh %F
Name=Default profile
```
`~/.local/bin/follow_symlink.sh`:
``` bash
#!/usr/bin/env bash

entries=""

for entry in "$@"; do
    if [ -h "$entry" ]; then
        entries="$entries,$(readlink "$entry")"
    fi
done

dbus-send \
    --session \
    --print-reply \
    --dest=org.freedesktop.FileManager1 \
    /org/freedesktop/FileManager1 \
    org.freedesktop.FileManager1.ShowItems \
    array:string:"${entries:1}" \
    string:""
```

Do you suggest any improvements to this?

Is it possible to enable this action only for symbolic links?

I have tried putting `MimeTypes=inode/symlink;` int the action profile, but it enable the action only for broken symlinks.

**Reply by tsujan** [reply]
> Do you suggest any improvements to this?

It seems good and should also work with all file managers that support `org.freedesktop.FileManager1` and custom actions, Thunar and Nautilus included (although I don't know whether GNOME devs have kept this feature in Nautilus — they usually remove features). Dolphin may work differently because KDE has its own tool called "service menu".

> I have tried putting MimeTypes=inode/symlink;

The mimetype of a symlink is that of its target. So, I'm afraid it'll appear with all types. You might want to check whether `readlink` returns an empty string and show a message if it does.

**Reply by tsujan** [reply]
BTW, if you only want to open the final targets of text-like symlinks, FeatherPad can already do it and also open and select the file with the active file manager (in its tab context menu).

**Comment by romildo** [comment]
> if you only want to open the final targets of text-like symlinks,

Most of the time I want to see other files in the same folder of the target file (or _near_ to it).
